vllm/model_executor/sampling_metadata.py

```diff
@@ -6,54 +6,269 @@
+        # prune the outcome logits from the model for the performance.
+        selected_token_indices: List[int] = []
+        # Used for selected_token_indices.
+        selected_token_start_idx = 0
+
+        # Sampling type -> (
+        # indices to sample within pruned output logits,
+        # indices to sample after running a triton sample kernel)
+        categorized_sample_indices: Dict[SamplingType,
+                                         List[Tuple[int, int]]] = {
+                                             t: []
+                                             for t in SamplingType
+                                         }
+        # Used for categorized_sample_indices.
+        categorized_sample_indices_start_idx = 0
+        categorized_sampled_token_indices_start_idx = 0
+
+        num_prompts = 0
+        for i, seq_group_metadata in enumerate(seq_group_metadata_list):
+            seq_ids = list(seq_group_metadata.seq_data.keys())
+            sampling_params = seq_group_metadata.sampling_params
+            is_prompt = seq_group_metadata.is_prompt
+            generator: Optional[torch.Generator] = None
+            # If the current seq group is in decode stage, it is None.
+            prompt_len: Optional[int] = None
+            subquery_len: Optional[int] = None
+            # prefill indices for this particular seq group.
+            prefill_indices = []
+            # sample indices for this particular seq group.
+            sample_indices = []
+            do_sample = seq_group_metadata.do_sample
+
+            if seq_group_metadata.is_prompt:
+                num_prompts += 1
+                assert len(seq_ids) == 1
+                assert subquery_lens is not None
+                assert prompt_lens is not None
+                subquery_len = subquery_lens[i]
+                prompt_len = prompt_lens[i]
+                num_samples = len(seq_ids)
+
+                # First, let's update what indices we want to post-process.
+                # The output logits will be pruned by selected_token_indices
+                # chosen from this logic.
+                if sampling_params.prompt_logprobs is not None:
+                    selected_token_end_idx = (selected_token_start_idx +
+                                              subquery_len)
+                    # If we need sampling, the last num_samples indexes are
+                    # for sampling.
+                    if do_sample:
+                        selected_token_end_idx -= num_samples
+
+                    # Select prompt indices for prompt logprob.
+                    selected_token_indices.extend(
+                        range(selected_token_start_idx,
+                              selected_token_end_idx))
 
-        self.num_prompts = len(prompt_lens) if prompt_lens is not None else 0
+                # Add sample indices for sampling and sample logprob if
+                # we need sampling.
+                if do_sample:
```

Comment:
This do_sample logic is ugly, but idk what's better way to do it. Comments are welcome